### DISCUSSION TITLE
Data Transfer Set additional headers

### DISCUSSION THREAD

**Discussion by galanisd** [eclipse-edc/Connector#3720] [Q&A] [answered]
I am working with EDC 0.2.1.

I initiate a data transfer from an HTTP endpoint to an other HTTP endpoint using the following msg.

```
{
  "@context": {
    "@vocab": "https://w3id.org/edc/v0.0.1/ns/"
  },
  "@type": "TransferRequest",
  "connectorId": "urn:connector:edc1",
  "connectorAddress": "http://localhost:8282/api/v1/ids",
  "contractId": "MQ==:YjYyM2U3NTUtNjIxMS00ZTI1LTk3ZTctMzM2MDI5OTBlZTI2:Zjc1MjQxN2QtNjA2OS00YTExLTk5MDEtYTViM2FmNTMwZDY4",                 
  "assetId": "b623e755-6211-4e25-97e7-33602990ee26",
  "protocol": "dataspace-protocol-http",
  "dataDestination": {
    "type": "HttpData",
    "baseUrl": "http://<SOME_IP>:1080/data4/zzz5.txt",
    "method": "PUT"
  }
}

```
I have set method to PUT (default is POST) using "method": "PUT" and it worked.
However I do not know how I can set additional headers.
The destination server requires Content-Length header:

`Error {411: Length Required} received writing HTTP data Asset Name to endpoint http://<SOME_IP>:1080/data4/zzz5.txt for request: Request{method=PUT, url=http://ldssetup.ilsp.gr:1080/data4/zzz5.txt}`

There is some documentation here: https://github.com/eclipse-edc/Connector/blob/v0.4.1/extensions/data-plane/data-plane-http/README.md?plain=1

but  I can't find an example on how to set Content-Length header (generally any additional header)

**Comment by galanisd** [comment]
I think that I found how to set an additional header

```
{
  "@context": {
    "@vocab": "https://w3id.org/edc/v0.0.1/ns/"
  },
  "@type": "TransferRequest",
  "connectorId": "urn:connector:edc1",
  "connectorAddress": "http://localhost:8282/api/v1/ids",
  "contractId": "MQ==:YjYyM2U3NTUtNjIxMS00ZTI1LTk3ZTctMzM2MDI5OTBlZTI2:Zjc1MjQxN2QtNjA2OS00YTExLTk5MDEtYTViM2FmNTMwZDY4",                 
  "assetId": "b623e755-6211-4e25-97e7-33602990ee26",
  "protocol": "dataspace-protocol-http",
  "dataDestination": {
    "type": "HttpData",
    "baseUrl": "http://<IP>:1080/data4/zzz5.txt",
    "method": "PUT",
    "header:Content-Length": "14"
  }  
}
```

The header parameter seems to reach HttpDataSink, it is printed in a log message (see below). However, I still get the same error

`SEVERE 2023-12-19T10:18:02.0339506 Error {411: Length Required} received writing HTTP data Asset Name to endpoint <IP>:1080/data4/zzz5.txt for request: Request{method=PUT, url=http://<IP>:1080/data4/zzz5.txt, headers=[Content-Length:14]}`

It looks like the header is never send from the httpclient or is somehow filtered

 ```
   @Override
    protected StreamResult<Void> transferParts(List<DataSource.Part> parts) {
        for (var part : parts) {
            var request = requestFactory.toRequest(params, part::openStream);
            try (var response = httpClient.execute(request)) {
                if (!response.isSuccessful()) {
                    monitor.severe(format("Error {%s: %s} received writing HTTP data %s to endpoint %s for request: %s",
                            response.code(), response.message(), part.name(), request.url().url(), request));
                    return ERROR_WRITING_DATA;
                }

                return StreamResult.success();
            } catch (Exception e) {
                monitor.severe(format("Error writing HTTP data %s to endpoint %s for request: %s", part.name(), request.url().url(), request), e);
                return ERROR_WRITING_DATA;
            }
        }
        return StreamResult.success();
    }
```

**Reply by ndr-brt** [reply, marked as answer]
By default, the http call is sent in a `chunked` manner, so it has no `Content-Length`.
If you want to change this, please add a `"nonChunkedTransfer": true` property to the `dataDestination` in that case the `Content-Length` header will be sent
